addons/website_sale/models/sale_order.py

```diff
@@ -129,9 +129,16 @@ def sale_product_domain(self, cr, uid, ids, context=None):
     def sale_get_order(self, cr, uid, ids, force_create=False, code=None, update_pricelist=None, context=None):
         sale_order_obj = self.pool['sale.order']
         sale_order_id = request.session.get('sale_order_id')
-        sale_order = sale_order_obj.browse(cr, SUPERUSER_ID, sale_order_id, context=context)
+        sale_order = None
+
+        # Test validity of the sale_order_id
+        if sale_order_id:
+            sale_order = sale_order_obj.browse(cr, SUPERUSER_ID, sale_order_id, context=context)
+            if not sale_order.exists():
+                sale_order_id = None
```

Comment:
Could use an old-API call for the exists and only set sale_order if the sale_order_id is valid as well, avoids having a possibly invalid sale_order recordset floating around:
```python
if sale_order_id and sale_order_obj.exists(cr, SUPERUSER_ID, sale_order_id, context=context):
    sale_order = sale_order_obj.browse(cr, SUPERUSER_ID, sale_order_id, context=context)
else:
    sale_order_id = None
```